include/exadg/fluid_structure_interaction_precice/coupling_interface.h

```diff
@@ -0,0 +1,224 @@
+#pragma once
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/vectorization.h>
+
+#include <deal.II/matrix_free/matrix_free.h>
+
+#include <exadg/matrix_free/integrators.h>
+#include <precice/SolverInterface.hpp>
+
+namespace Adapter
+{
+using namespace dealii;
+
+/**
+ * Enum to handle all implemented data write methods one can use
+ */
+enum class WriteDataType
+{
+  undefined,
+  values_on_dofs,
+  values_on_other_mesh,
+  gradients_on_other_mesh,
+  values_on_quads,
+  normal_gradients_on_quads
+};
+
+/**
+ * A pure abstract base class, which defines the interface for the functions
+ * used in the main Adapter class. Each instance of all derived classes are
+ * always dedicated to a specific coupling mesh and may provide functions on
+ * how to read and write data on this mesh and how to define the mesh.
+ */
+template<int dim, int data_dim, typename VectorizedArrayType>
+class CouplingInterface
+{
+public:
+  CouplingInterface(std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> data,
+                    std::shared_ptr<precice::SolverInterface>                           precice,
+                    std::string                                                         mesh_name,
+                    types::boundary_id interface_id);
+
+  virtual ~CouplingInterface() = default;
+
+  /// Alias for the face integrator
+  using FEFaceIntegrator = FaceIntegrator<dim, data_dim, double, VectorizedArrayType>;
```

Comment:
@peterrum With dependencies on `MatrixFree` and `FEFaceEvaluation`, this class is probably not implemented in a dealii-general-purpose way as you have written in your comment?